pkg/reconciler/apis/permissionclaimlabel/permissionclaimlabel_reconcile.go

```diff
@@ -0,0 +1,242 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package permissionclaimlabel
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	aggregateerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/informers"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/apis/conditions/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/util/conditions"
+	"github.com/kcp-dev/kcp/pkg/indexers"
+)
+
+// reconcilePermissionClaims determines the resources that need to be labeled for access by a permission claim.
+// It determines what permissions need to be added, what permissions need to be removed.
+// It also updates the status if it finds an invalid permission claim.
+// Permission claims are considered invalid when the identity hashes are mismatched, and when there is no dynamic informer
+// for the group resource.
+func (c *controller) reconcile(ctx context.Context, apiBinding *apisv1alpha1.APIBinding) error {
+	logger := klog.FromContext(ctx)
+
+	clusterName := logicalcluster.From(apiBinding)
+
+	if apiBinding.Status.BoundAPIExport == nil {
+		return nil
+	}
+
+	exportClusterName := apiBinding.Status.BoundAPIExport.Workspace.Path
+	exportName := apiBinding.Status.BoundAPIExport.Workspace.ExportName
+	export, err := c.getAPIExport(logicalcluster.New(exportClusterName), exportName)
+	if err != nil {
+		logger.Error(err, "error getting APIExport", "exportClusterName", exportClusterName, "exportName", exportName)
+		return nil // nothing we can do
+	}
+
+	logger = logger.WithValues("apiExportWorkspace", logicalcluster.From(export), "apiExportName", export.Name)
```

Comment:
logging.From, no?

Reply:
that's already in for the apiBinding

Reply:
`logging.WithObject(logger, export)`?

Reply:
Fixed (`logger = logger.WithValues(logging.FromPrefix("apiExport", apiExport))`)

Reply:
oh, thx